### ISSUE TITLE
hardsigmoid replacement

### ISSUE THREAD

**Issue by livingbeams** [PeterL1n/RobustVideoMatting#47]
I've been trying to export an onnx model replacing the hardsigmoid operator.

I have modified the site-packages/torch/onnx/symbolic_opset9.py file this way:

@parse_args("v")
def hardswish(g, self):
    hardsigmoid = g.op('HardSigmoid', self, alpha_f=1 / 6)
    return g.op("Mul", self, hardsigmoid)

@parse_args("v")
def hardsigmoid(g, self):
    hardsigmoid = g.op('HardSigmoid', self, alpha_f=1 / 6)
    return g.op("Mul", self, hardsigmoid) 

But I am not sure at all if this is the way to replace them with primitive ops

When I export the onnx with this change I still get and error "OnnxImportException: Unknown type HardSigmoid encountered while parsing layer 396" with the inference engine I am trying to use.

**Comment by PeterL1n** [comment]
So your inference framework does not support `HardSigmoid`. But no worries, hard sigmoid can be replaced with more primitives ops.

You can change them to formula below. The actual export code you need to figure out. Shouldn't be difficult.
```
Hardsigmoid(x) = x.clamp(0, 6).add(3).div(6).
Hardswish(x) = x * Hardsigmoid(x)
```

**Comment by livingbeams** [comment]
Hi,

I've been trying to figure out how to replace the HardSigmoid activation with no luck.

Do you think it could be done overwriting
@parse_args('v')
def hardsigmoid(g, self):
    return g.op('HardSigmoid', self, alpha_f=1 / 6)
in the file \site-packages\torch\onnx\symbolic_opset9.py

Or do you think it would be necessary to modify decoder.py and lrsapp.py where it uses nn.Sigmoid() ?

Regards.

**Comment by livingbeams** [comment]
What I have tried is this:

@parse_args('v')
def hardsigmoid(g, self):
    min_val = g.op('Constant', value_t=torch.tensor(0, dtype=torch.float))
    max_val = g.op('Constant', value_t=torch.tensor(6, dtype=torch.float))
    clip = g.op("Clip", self, min_val, max_val)
    add = g.op("Add", clip, g.op('Constant', value_t=torch.tensor(3, dtype=torch.float)))
    div = g.op("Div", add, g.op('Constant', value_t=torch.tensor(6, dtype=torch.float)))
    return div

**Comment by livingbeams** [comment]
I've found out the correct order of the operators is:

@parse_args('v')
def hardsigmoid(g, self):
    add = g.op("Add", self, g.op('Constant', value_t=torch.tensor(3, dtype=torch.float)))
    clip = g.op("Clip", add, g.op('Constant', value_t=torch.tensor(0, dtype=torch.float)), g.op('Constant', value_t=torch.tensor(6, dtype=torch.float)))
    div = g.op("Div", clip, g.op('Constant', value_t=torch.tensor(6, dtype=torch.float)))
    return div

This way we can avoid the use of hardsigmoid operator

The pity is that it stills doesn't work with barracuda unity package. I gives an error:
"Must have an input rank for 804 in order to convert axis for NHWC op"

**Comment by PeterL1n** [comment]
Can you use Neutron (https://github.com/lutzroeder/Netron) to open the ONNX file you exported and find out what exactly is node 804?

**Comment by livingbeams** [comment]
Sure! 
It is the output of a Slice block (starts <2>, axes <2>) to a Concat block.

![image](https://user-images.githubusercontent.com/8204316/136333310-43e08f4f-58c6-4dc7-9dd0-38ce5e210981.png)

**Comment by PeterL1n** [comment]
Here is a hack you can try.

Change this line
https://github.com/PeterL1n/RobustVideoMatting/blob/ba3791548b2505ac185de2ab6b075d54b6bdbc69/model/decoder.py#L83
to
```py
x = F.interpolate(x, s.shape[2:], mode='bilinear', align_corners=False)
```

And then remove the following line. 
https://github.com/PeterL1n/RobustVideoMatting/blob/ba3791548b2505ac185de2ab6b075d54b6bdbc69/model/decoder.py#L84

This will just remove the slice operation altogether. It will create some subtle differences compared to how the model was trained, but I think it is minor enough for most use cases.

**Comment by DefTruth** [comment]
> So your inference framework does not support `HardSigmoid`. But no worries, hard sigmoid can be replaced with more primitives ops.
> 
> You can change them to formula below. The actual export code you need to figure out. Shouldn't be difficult.
> 
> ```
> Hardsigmoid(x) = x.clamp(0, 6).add(3).div(6).
> Hardswish(x) = x * Hardsigmoid(x)
> ```

it seems that should change  `Hardsigmoid` to `x.add(3).clamp(0,6).div(6)`, not `x.clamp(0, 6).add(3).div(6)`:
```python
import torch
from torch.nn import functional as F

if __name__ == "__main__":
    x = torch.FloatTensor([-4, -3. -2., -1., 0., 1., 2., 3., 4.])
    hs1 = x.add(3).clamp(0, 6).div(6)
    hs2 = F.hardsigmoid(x)
    rest = hs1 - hs2
    print(hs1)
    print(hs2)
    print(rest)
```  
log:  
```shell
tensor([0.0000, 0.0000, 0.3333, 0.5000, 0.6667, 0.8333, 1.0000, 1.0000])
tensor([0.0000, 0.0000, 0.3333, 0.5000, 0.6667, 0.8333, 1.0000, 1.0000])
tensor([0., 0., 0., 0., 0., 0., 0., 0.])
```  
the result of RVM is not right if i change  `Hardsigmoid` to `x.clamp(0, 6).add(3).div(6)`, but `x.add(3).clamp(0,6).div(6)` work fined. the official implementation of  F.hardsigmoid in pytorch is:  
![image](https://user-images.githubusercontent.com/31974251/136692083-28d19158-d440-4076-9eae-f669489ea08a.png)
i think we should add(3) first and then clamp(0, 6) with post div(6), it make x -> (x+3):  
> 0, if (x+3)<=-3+3=0
> 1, if (x+3)>=+3+3=6
> (x+3)/6, otherwise

**Comment by PeterL1n** [comment]
Yeah. That’s probably my typo.

**Comment by livingbeams** [comment]
Hi again,

I've tried the hack you have proposed to replace the upsampling with:
x = F.interpolate(x, s.shape[2:], mode='bilinear', align_corners=False)

The inference seems to be pretty similar but now there is a similar error at node 818
"Must have an input rank for 818 in order to convert axis for NHWC op"

![image](https://user-images.githubusercontent.com/8204316/137079698-1c555383-a3b1-43dc-b5d3-d6c7ee5c4822.png)

It is the output of the slice (starts 2, axes 2) to the concat operator.

I also get a bunch of warnings and I am not sure if these are critical to make the inference work

![image](https://user-images.githubusercontent.com/8204316/137080272-7c33dfe9-5a21-4046-b7cb-a3bf71b4c9fa.png)

![image](https://user-images.githubusercontent.com/8204316/137080744-8c676781-6894-4cd6-a587-1d76ac7fd008.png)

Best regards

**Comment by PeterL1n** [comment]
Change this too:
https://github.com/PeterL1n/RobustVideoMatting/blob/3dc6b6a4990fa4895b17afb817a26847450886c8/model/decoder.py#L128

**Comment by livingbeams** [comment]
This time I still get the same error:
"Must have an input rank for 818 in order to convert axis for NHWC op"

I've tried to replace the other uses of upsample:

    def forward_single_frame(self, x, f, s, r):
        #x = self.upsample(x)
        x = F.interpolate(x, s.shape[2:], mode='bilinear', align_corners=False)

    def forward_time_series(self, x, f, s, r):
        B, T, _, H, W = s.shape
        x = x.flatten(0, 1)
        f = f.flatten(0, 1)
        s = s.flatten(0, 1)
        #x = self.upsample(x)
        x = F.interpolate(x, s.shape[2:], mode='bilinear', align_corners=False)

    def forward_single_frame(self, x, s):
        #x = self.upsample(x)
        x = F.interpolate(x, s.shape[2:], mode='bilinear', align_corners=False)

    def forward_time_series(self, x, s):
        B, T, _, H, W = s.shape
        x = x.flatten(0, 1)
        s = s.flatten(0, 1)
        #x = self.upsample(x)
        x = F.interpolate(x, s.shape[2:], mode='bilinear', align_corners=False)

But I still get the same error.

**Comment by PeterL1n** [comment]
What’s 818 now

**Comment by livingbeams** [comment]
It is the output of slice(starts 2, axes 2) to the input of concat just before introducing the recurrent state r3i in the network:

![image](https://user-images.githubusercontent.com/8204316/138317696-886f6d01-edd9-435a-841f-d0c0def01aa8.png)


I greatly appreciate the time spent suggesting the changes, but I am afraid that it will not be possible to overcome all the current limitations of inference with Barracuda.

**Comment by leavittx** [comment]
Still trying to get it working.

Did replacements  above:
![image](https://user-images.githubusercontent.com/33001/158967387-9addf426-23a1-4cb2-9c91-f06d516d6dce.png)

decoder.py changes:
![image](https://user-images.githubusercontent.com/33001/158968910-73902f57-bf0b-49a2-8398-aa6cf7555e85.png)
![image](https://user-images.githubusercontent.com/33001/158969167-90ae1ecb-72d9-467c-a397-f0bf545c735d.png)

Removed `dynamic_axes `parameter of `torch.onnx.export`, as it's not fully supported.

Current Barracuda error on model import:
`Must have input rank for 671 in order to convert axis for NHWC op`
Netron screenshot for 671:
![image](https://user-images.githubusercontent.com/33001/158971759-0a32d447-bd47-413e-a78e-efa57988ef9c.png)

[Model](https://www.dropbox.com/s/1g34ot9uw0v3ulu/robustvideomatting_model_barracuda_v01.onnx?dl=0)
Export cmd: `python export_onnx.py  --model-variant mobilenetv3  --checkpoint work\checkpoint\rvm_mobilenetv3.pth --precision float32  --opset 12  --device cuda --output model.onnx`

Any help highly appreciated
